proposals/0000-single-pattern-checks.md

```diff
@@ -0,0 +1,106 @@
+# Single pattern checks
+
+* Proposal: [HXP-NNNN](NNNN-single-pattern-checks.md)
+* Author: [Dmitrii Maganov](https://github.com/vonagam)
+
+## Introduction
+
+New syntax for single pattern checks.
+
+## Motivation
+
+Current single pattern check - `match` method on an enum value - is limited in functionality compared to usual pattern matching. No captures or guards and only a single enum value as a root value.
+
+Proposed syntax will allow to use full arsenal of matcher for three common scenarios.
+
+## Detailed design
+
+All of single pattern checks allias such switch:
+
+```haxe
+switch ($value) {
+  case $pattern if ($guard): $ifBody;
+  case _: $elseBody;
+}
+```
+
+### Pattern test with `is case`
+
+```
+$value is case $pattern [where ($guard)]
+```
+
+With `$ifBody` being `true` and `$elseBody` - `false`.
+
+```haxe
+function isSome(option: Option<Int>): Bool {
+  return option is case Some(_);
+}
+
+function isDivisibleSome(option: Option<Int>, divisor: Int): Bool {
+  return option is case Some(int) where (int % divisor == 0);
+}
+```
+
+### Pattern condition with `if case`
+
+```
+if case ($value, $pattern[, $guard]) $ifBody [else $elseBody]
```

Comment:
Why not `if case ($value, $pattern[ where $guard]) $ifBody [else $elseBody]` instead, for consistency? `$guard` separated with a comma doesn't look so obvious to me.

Reply:
For full consistency it can be `if case ($value is $pattern where $guard)`. 
Right now `is` is a comma and so is `where`.
I am okay with either variant, went for shorter one.

Reply:
Yeah, `is` would be better too imo

Reply:
`is` alone (without `case`) works ok for enum values, but less for other types:
`if case (player is {health: 0}) {}`
Don't know, doesn't feel right.

If the same word had to be used in test and in condition then i would rename `is case` to `match case` and use `if case ($value match $pattern)`.

Reply:
Updated, current version does not have such inconsistencies.